app/stac_api/models.py

```diff
@@ -494,6 +492,37 @@ def upload_asset_to_path_hook(instance, filename=None):
     return get_asset_path(instance.item, instance.name)
 
 
+class DynamicStorageFileField(models.FileField):
+
+    def pre_save(self, model_instance: "Asset", add):
+        """Determine the storage to use for this file
+
+        The storage is determined by the collection's name. See
+        settings.MANAGED_BUCKET_COLLECTION_PATTERNS
+        """
+        collection = model_instance.item.collection
+
+        bucket = select_s3_bucket(collection.name).name
+
+        parts = settings.STORAGES[bucket]['BACKEND'].split(".")
+        storage_module_name = ".".join(parts[:-1])
+        storage_cls_name = parts[-1:]
+        storage_module = __import__(storage_module_name, fromlist=[parts[-2:-1]])
+        storage_cls = getattr(storage_module, storage_cls_name[0])
+
+        self.storage = storage_cls()
+
+        instance = model_instance
+
+        # we need to specify the storage for the actual
+        # file as well
+        instance.file.storage = self.storage
```

Comment:
Oh this is somewhat a leftover from changing the code all the time. This would make more sense as
`model_instance.file.storage = self.storage`
What this does is set the correct file storage on the actual file, so everything that uses the storage on the file instance directly will have the correct one.
Also see `upload_asset_to_path_hook` in this file and `BaseS3Storage.get_object_parameters` for how they interact across multiple code locations